wpengine/wp-graphql-content-blocks
# fix: cleanup constants and refactor autoload handling

##  What

This PR fixes the plugin's naive autoload detection which previously would:
1. throw a fatal error if dependencies were not installed (instead of the intended `admin_notice`)
2. show an `admin_notice` if the plugin was installed as a composer dependency itself (even though the plugin dependencies would get installed).

## How

1. Autoload handling was moved to `\WPGraphQL\ContentBlocks\Autoloader`, where it checks (in order):
  a. if `WPGRAPHQL_CONTENT_BLOCKS_AUTOLOAD` wants us to skip autoloading
  b. if we've previously loaded the dependencies (via a static property)
  c. if the `WPGraphQLContentBlocks` class exists (that means it's being autoloaded already)
  d. and only _then_ if there is a local `vendor/autoload.php` file.
2. The autoloader now runs in `wp-graphql-content-blocks.php::wpgraphql_content_blocks_init()`, where it is `required_once`ed instead of the (now autoloaded) `WPGraphQLContentBlocks` class.
3. To support this, the several constants that were defined in `WPGraphQLContentBlocks` have been moved to the pluggable `wp-graphql-content-blocks.php:wpgraphql_content_blocks_constants()`, (The versioning script was updated to accomodate the change)
4. Since the constants were already being merged to a single source, I took a moment to dedupe and deprecate the old ones:
  - `WPGRAPHQL_CONTENT_BLOCKS_FILE` => `WPGRAPHQL_CONTENT_BLOCKS_PLUGIN_FILE`
  - `WPGRAPHQL_CONTENT_BLOCKS_PATH` => `WPGRAPHQL_CONTENT_BLOCKS_PLUGIN_PATH`
  - `WPGRAPHQL_CONTENT_BLOCKS_MIN_PHP_VERSION` => no replacement (this seems to be an artifact from wp-graphql core boilerplate).
  
## To test

1. Create a composer-based WordPress install (e.g. bedrock)
2. Add the branch to the `composer.json` repositories:
```json
"repositories": [
  {
    "type": "vcs",
    "url": "https://github.com/rtcamp/wp-graphql-content-blocks.git"
  },
],
```
3. Install the plugins as composer dependencies
```bash
  composer require wpackagist-plugin/wp-graphql wpengine/wp-graphql-content-blocks:dev-fix/constants-autoloading
```

4. Go to the Plugins Page, activate the plugin, see everything works with no missing deps admin notice.

  
## Additional notes
  `@todo` tags are used for version numbers and should be replaced once we know what the version will be.
